momaland/learning/continuous_ppo/momappo_crazyrl.py

```diff
@@ -0,0 +1,550 @@
+    single_obs_space = env.observation_space(env.possible_agents[0])
+
+    actor = Actor(single_action_space.shape[0], activation=args.activation)
+    critic = Critic(activation=args.activation)
+    key, actor_key, critic_key = jax.random.split(key, 3)
+    dummy_local_obs_and_id = jnp.zeros(single_obs_space.shape)
+    dummy_global_obs = jnp.zeros(env.state().shape)
+    actor_params = actor.init(actor_key, dummy_local_obs_and_id)
+    critic_params = critic.init(critic_key, dummy_global_obs)
+    if args.anneal_lr:
+        tx = optax.chain(
+            optax.clip_by_global_norm(args.max_grad_norm),
+            optax.adam(learning_rate=linear_schedule, eps=1e-5),
+        )
+    else:
+        tx = optax.chain(
+            optax.clip_by_global_norm(args.max_grad_norm),
+            optax.adam(args.lr, eps=1e-5),
+        )
+
+    actor_train_state = TrainState.create(
+        apply_fn=actor.apply,
+        params=actor_params,
+        tx=tx,
+    )
+
+    critic_train_state = TrainState.create(
+        apply_fn=critic.apply,
+        params=critic_params,
+        tx=tx,
+    )
+
+    # BUFFER
+    buffer = Buffer(
+        batch_size=args.num_steps,
+        joint_actions_shape=(len(env.possible_agents), single_action_space.shape[0]),
+        obs_shape=(len(env.possible_agents), single_obs_space.shape[0]),
+        global_obs_shape=env.state().shape,
+        num_agents=len(env.possible_agents),
+    )
+
+    def _to_array_obs(obs: dict):
+        """Converts a dict of observations to a numpy array of shape (num_agents, obs_dim)."""
+        return np.stack([obs[agent] for agent in env.possible_agents])
+
+    @jax.jit
+    def _ma_get_pi(params, obs: jnp.ndarray):
+        """Gets the actions for all agents at once. This is done with a for loop because distrax does not like vmapping."""
+        return [actor.apply(params, obs[i]) for i in range(len(env.possible_agents))]
+
+    def _batched_ma_get_pi(params, obs: jnp.ndarray):
+        """Gets the actions for all agents in all the envs at once. This is done with a for loop because distrax does not like vmapping."""
+        return [actor.apply(params, obs[:, i, :]) for i in range(len(env.possible_agents))]
+
+    @jax.jit
+    def _ma_sample_and_log_prob_from_pi(pi: List[MultivariateNormalDiag], key: chex.PRNGKey):
+        """Samples actions for all agents in all the envs at once. This is done with a for loop because distrax does not like vmapping.
+
+        Args:
+            pi (List[MultivariateNormalDiag]): List of distrax distributions for agent actions (batched over envs).
+            key (chex.PRNGKey): PRNGKey to use for sampling: size should be (num_agents, 2).
+        """
+        return [pi[i].sample_and_log_prob(seed=key[i]) for i in range(len(env.possible_agents))]
+
+    # Batch get value
+    vmapped_get_value = vmap(critic.apply, in_axes=(None, 0))
+
+    critic.apply = jax.jit(critic.apply)
+
+    @jax.jit
+    def _calculate_gae(traj_batch, last_val):
```

Comment:
Did you change something over here compared to the original?

Reply:
No, the only thing I changed is `num_drones` to `len(env.possible_agents)`.